lib/src/kvs/ds.rs

```diff
@@ -473,6 +479,38 @@ impl Datastore {
 	// End cluster helpers, storage functions here
 	// -----
 
+	/// Use for internal database operations so they are retriable.
+	/// This is so that we have a centralised implementation for handling retries
+	/// for potentially conflicting operations.
+	///
+	/// This also handles transactions always being closed by forgetting to commit.
+	///
+	/// Do not use for user operations such as queries or status requests.
+	pub async fn configured_transaction<F, G, T>(
+		&self,
+		config: Config,
+		write: bool,
+		lock: bool,
+		retriable: F,
+	) -> Result<G, Error>
+	where
+		F: Fn(&mut Transaction) -> Future<Output = T> + Send + 'static,
+		T: Send + 'static,
+	{
+		let mut last_err = None;
+		for _ in 0..config.retries {
+			let mut tx = self.transaction(write, lock).await?;
+			let res = retriable(&mut tx);
+			if res.is_ok() {
+				tx.commit().await?;
```

Comment:
If the transaction fails to commit due to a conflict, shouldn't we retry rather than returning?

Reply:
Oh yeah, you are right! Thanks